vmgr/vmgrmodifyweight.man
.\" @(#)$RCSfile: vmgrmodifyweight.man,v $ $Revision: 1.1 $ $Date: 2005/03/17 10:23:36 $ CERN IT-PDP/DM Jean-Philippe Baud
.\" Copyright (C) 2000-2002 by CERN/IT/PDP/DM
.\" All rights reserved
.\"
.TH VMGRMODIFYWEIGHT 1 "$Date: 2005/03/17 10:23:36 $" CASTOR "vmgr Administrator Commands"
.SH NAME
vmgrmodifyweight \- modify the weight and/or delta assigned to a given DGN in the CASTOR Volume Manager
.SH SYNOPSIS
.B vmgrmodifyweight
.BI --dgn " dgn"
[
.BI --weight " weight"
|
.BI --delta " delta"
]
.SH DESCRIPTION
.B vmgrmodifyweight
modifies the weight and/or delta assigned to a given DGN in the CASTOR Volume Manager
.TP
.I dgn
specifies the dgn we want to modify the weight of
.TP
.I weight
is the numeric weight we want to assign to the dgn
.TP
.I delta
is the numeric delta we want to assign to the dgn
.LP
The dgn parameter is mandatory, and either weight or delta should appear
(or both of them). When a tape is being selected for writing in vmgr
(via vmgr_gettape), the weight of a given DGN determines which 
one is going to be chosen; higher values for the weight give priority
to that dgn. For each selection, the weigth will be increased the amount
specified per delta.
.LP
This command requires ADMIN privilege in the Cupv database.
.SH EXIT STATUS
This program returns 0 if the operation was successful or >0 if the operation
failed.
.SH SEE ALSO
.BR Castor_limits(4) ,
.BR Cupvlist(1) ,
.B vmgr_modifyweight(3)
.SH AUTHOR
\fBCASTOR\fP Team <castor.support@cern.ch>


CHANGELOG.md
## 5.11.23.0-1 (2026-09-25)

### New Features

- [frontend] Add user configuration improvements for cta-admin-grpc (cta/CTA!1635)

## 5.11.22.0-1 (2026-09-04)

### New Features

- [frontend] Implement JWT blacklisting on the CTA Frontend (cta/CTA!1548)

- [runtime] Add additional validation logic to application config (cta/CTA!1567)


### Bug Fixes

- [scheduler] Fix issues after pg1 release rest: repack admin commands, archive...

- [Tools] Fix 'priority' column in 'cta-admin dr ls' command (cta/CTA!1546)

- [runtime] Make signals, telemetry and health handling more robust (cta/CTA!1554)

- [runtime] Add bounds checking and better error diagnostics to TOML loading (cta/CTA!1559)

## 5.11.21.0-1 (2026-08-25)

### New Features

- [scheduler] Add scheduler workflow metrics to OpenTelementry (cta/CTA!1387)


### Changes

- [Tools] Move cta-admin af ch command to tf ch (cta/CTA!1525)
- [Tools] Replace cta-objectstore-tools.conf with cta-scheduler.conf (cta/CTA!1540)
- [maintd] Change scheduler.objectstore_backend_path to scheduler.config_file (cta/CTA!1518)


### Bug Fixes

- [frontend] Fix logic bug in frontend authentication and SEGFAULT in cta-admin (cta/CTA!1558)
- [Tools] Fix cta-verify-file (cta/CTA!1545)
- [Tools] Add `--storage-class` option to `cta-send-event` (cta/CTA!1542)


### Performance Improvements

- [taped] Skip block reposition request if drive is already positioned correctly (cta/CTA!1527)

## 5.11.20.0-1

### New Features

- [Tools] Add optional storage class statistics, per tape, to cta-admin tool (cta/CTA!1410)
- [Misc] Add event_name to important CTA log messages (cta/CTA!1381)

### Changes

- [Tools] Change short option for cta-admin af ch --idfile from -I to -F (cta/CTA!1453)
- [taped] Taped no longer sets its own user, group and SYS_RAWIO capability (cta/CTA!1426)
- [Misc] Improve .service and .logrotate configuration (cta/CTA!1524)
- [Misc] Modernise RPM spec metadata and file handling (cta/CTA!1507)
- [Misc] Move systemd units to `/usr/lib/systemd/system` (cta/CTA!1497)
- [Misc] Change systemd to weak RPM dependency of CTA (cta/CTA!1509)
- [Misc] Change logrotate to weak RPM dependency of CTA (cta/CTA!1505)
- [Misc] Remove RPM changelog; package CHANGELOG.md as doc (cta/CTA!1508)
- [Misc] Remove noreplace from example configuration files (cta/CTA!1506)
- [taped] Drive process crashing produces CRIT message instead of INFO (cta/CTA!1499)
- [taped] cta-eosdf no longer sources sysconfig (cta/CTA!1455)

### Bug Fixes

- [frontend] Forbid reclaiming tapes with recently deleted files (cta/CTA!1418)
- [frontend] PREPARE gRPC request shouldn't fail without a storage class (cta/CTA!1471)
- [taped] Fix crash due to accidental pointer arithmetic in exception msg (cta/CTA!1502)
- [taped] Fix opentelemetry shutdown race condition with objectstore (cta/CTA!1500)
- [scheduler]: Fix archive unmount decision logic for postgres scheduler (cta/CTA!1450)
- [frontend] Fix frontend catalogue cache for mount rules (cta/CTA!1490)
- [CI] Ensure CTA images trust CERN Root CA (cta/CTA!1535)
- [Misc] Ensure CMake correctly installs executables as PROGRAMS (cta/CTA!1523)

### Security

- [CI] Generate SBOM for each Docker image (cta/CTA!1532)

### Other

- [CI] Add system tests to cover staging tokens on EOS-CTA (cta/CTA!1493)

## 5.11.19.0-1

### Changes

- [frontend] Split Admin/WFE frontends, add mTLS (cta/CTA!1367)
- [catalogue] Forbid adding or changing encryption key of non-empty tapes (cta/CTA!1414)

### Deprecations

- [frontend] Deprecate storage class field in EOS-CTA xattrs (cta/CTA!1424)

### Other

- [CI] Add Enstore and EnstoreLarge tape tests (cta/CTA!1412)

## 5.11.18.0-2

### Bug Fixes

- [Misc] Update required runtime version of opentelemetry-cpp to v1.26.0 (cta/CTA!1370)

## 5.11.18.0-1

### New Features

- [maintd] Add catalogue and telemetry conf files to runtime directory (cta/CTA!1303)
- [Misc] Package basic JSON logging schema (cta/CTA!1304)
- [Misc] Add source location and CTA version to log lines (cta/CTA!1348)

### Changes

- [frontend] Allow admin commands or physics workflows to be disabled on CTA frontend (cta/CTA!1336)
- [Misc] Replace reflect-cpp library with custom implementation (cta/CTA!1291)
- [Misc] Change Catalogue backend install behaviour (cta/CTA!1295)

### Bug Fixes

- [catalogue] Improve CTA resilience to catalogue DB outages during data fetching (cta/CTA!1334)
- [taped] Improve readExactBlock error message contents (cta/CTA!1311)
- [taped] Enabling reading of tapes with misformed OSM tape format (cta/CTA!1343)
- [frontend] Prevent 'cta-admin dr rm' command from crashing the cta-frontend (cta/CTA!1305)
- [frontend] Remove race conditions in NegotiationRequestHandler (cta/CTA!1310)
- [frontend] Fix broken cta-admin fr ls command (cta/CTA!1340)
- [Misc] Fix compile issues when building CTA in Release mode (cta/CTA!1207)

### Performance Improvements

- [CI] Use python xrootd persistent clients for archive instead of xrdcp --parallel (cta/CTA!1274)

### Other

- [Misc] Update opentelemetry-cpp to v1.26.0 (cta/CTA!1321)


## 5.11.17.0-1

### New Features

- [maintd] Add CLI option to use runtime directory for app metadata (cta/CTA!1266)

### Changes

- [maintd] Update maintd to use TOML configuration (cta/CTA!1246)
- [maintd] Use declarative config to initialise OpenTelemetry SDK (cta/CTA!1248)
- [maintd] Add use_unix_domain_socket configuration entry for health server (cta/CTA!1282)
- [maintd] Remove daemon_user and daemon_group settings (cta/CTA!1252)

### Bug Fixes

- [frontend] Fix crash in 'cta-admin dr ls' command when using multiple...
- [frontend] Add error message when trying to change media type of non-empty tape (cta/CTA!1253)

### Removals

- [frontend] Stop using TAPE_POOL.SUPPLY column (cta/CTA!1245)

### Performance Improvements

- [taped] Optimise DriveState queries (cta/CTA!1198)

### Other

- [Misc] Rename *.conf.example files to *.example.conf (cta/CTA!1279)

## 5.11.16.1-1

### Bug Fixes


vmgr/vmgrmodifyweight.c
/*
 * Copyright (C) 2000-2002 by CERN/IT/PDP/DM
 * All rights reserved
 */

#ifndef lint
static char sccsid[] = "@(#)$RCSfile: vmgrmodifyweight.c,v $ $Revision: 1.1 $ $Date: 2005/03/17 10:23:36 $ CERN IT-PDP/DM Jean-Philippe Baud";
#endif /* not lint */

/*	vmgrmodifyweight - modify the current weight of a given dgn */
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include "Cgetopt.h"
#include "serrno.h"
#include "vmgr_api.h"
main(argc, argv)
int argc;
char **argv;
{
	int c;
        char *dp;
	int errflg = 0;
	char *dgn = NULL;
	static struct Coptions longopts[] = {
		{"dgn", REQUIRED_ARGUMENT, 0, OPT_DGN },
		{"weight", REQUIRED_ARGUMENT, 0, OPT_WEIGHT},
		{"delta", REQUIRED_ARGUMENT, 0, OPT_DELTA},
		{0, 0, 0, 0}
	};
	int weight = -1;
	int delta = 0;
	int valid = 0;

	Copterr = 1;
	Coptind = 1;
        while ((c = Cgetopt_long (argc, argv, "d:w:", longopts, NULL)) != EOF) {
                switch (c) {
		case 'd':
		case OPT_DGN:
			dgn = Coptarg;
			break;
		case 'w':
		case OPT_WEIGHT:
                        if ((weight = strtol (Coptarg, &dp, 10)) < 0 ||
                                    *dp != '\0') {

				fprintf (stderr,
				   "invalid weight%s\n", Coptarg);
				errflg++;
			}
			valid = valid | OPT_VALID_WEIGHT;
                        break;
                case OPT_DELTA:
                        if ((delta = strtol (Coptarg, &dp, 10)) < 0 ||
                                    *dp != '\0') {
                                                                                
                                fprintf (stderr,
                                   "invalid delta%s\n", Coptarg);
                                errflg++;
                        }
			valid = valid | OPT_VALID_DELTA;
                        break;

                default:
			printf("Undefined option\n");
                        break;
                }
        }
        if (Coptind < argc) {
                errflg++;
        }
        /* DGN arguments is mandatory, and either weight, delta, or both
           should also be provided. Check that they are passed */
        if ((dgn == NULL) || (valid == 0)) {
		errflg++;
        }
        if (errflg) {
                fprintf (stderr, "usage: %s %s", argv[0],
		    "--dgn dgn [--weight weight | --delta delta]\n");
                exit (USERR);
        }
 
        /* valid specifies wich parameters have valid data */
	if (vmgr_modifyweight (dgn, weight, delta, valid) < 0) {
		fprintf (stderr, "vmgrmodifyweight %s: %s\n", dgn, 
		    (serrno == ENOENT) ? "No such dgn" : sstrerror(serrno));
		exit (USERR);
	}
	exit (0);
}
